File: trunk/Bricksmith/UnitTests/Classes/LDraw/Commands/LPubRemoveGroup_Tests.m
//
//  LPubRemoveGroup_Tests.m
//  UnitTests
//
//  Created by Sergey Slobodenyuk on 2023-03-09.
//

#import "LPubRemoveGroup.h"

#import <XCTest/XCTest.h>
#import "MockArchiver.h"
#import "MockScanner.h"


// MARK: Mocks -

@interface LPubRemoveGroup ()

+ (LPubCommand *) lpubCommandInstance:(NSArray<NSString *> *)parameters;

@end


// MARK: - Tests -

@interface LPubRemoveGroup_Tests : XCTestCase

@end

@implementation LPubRemoveGroup_Tests

- (void)test_LPubRemoveGroup_EncodedProperly
{
	MockArchiver *encoder = [MockArchiver new];
	LPubRemoveGroup *command = [[LPubRemoveGroup alloc] init];
	command.groupName = @"Some string";

	[command encodeWithCoder:encoder];

	XCTAssertEqual(encoder.data.count, 3);
	XCTAssertEqualObjects(encoder.data[@"groupName"], @"Some string");
	XCTAssertEqualObjects(encoder.data[@"lpubCommandString"], @"REMOVE GROUP \"Some string\"");
	XCTAssertEqualObjects(encoder.data[@"commandString"], @"!LPUB REMOVE GROUP \"Some string\"");
}


- (void)test_LPubRemoveGroup_DecodedProperly
{
	MockArchiver *decoder = [MockArchiver new];
	decoder.data = [@{
				@"groupName" : @"Some string",
		@"lpubCommandString" : @"REMOVE GROUP \"Some string\"",
			@"commandString" : @"!LPUB REMOVE GROUP \"Some string\""
	} mutableCopy];
	
	LPubRemoveGroup *command = [[LPubRemoveGroup alloc] initWithCoder:decoder];
	command.groupName = @"Some string";

	XCTAssertEqualObjects(command.groupName, @"Some string");
	XCTAssertEqualObjects(command.lPubCommandString, @"REMOVE GROUP \"Some string\"");
	XCTAssertEqualObjects(command.commandString, @"!LPUB REMOVE GROUP \"Some string\"");
}


- (void)test_LPubRemoveGroup_CopiedProperly
{
	LPubRemoveGroup *command = [[LPubRemoveGroup alloc] init];
	command.groupName = @"Some string";

	LPubRemoveGroup *duplicate = [command copyWithZone:nil];
	
	XCTAssertEqualObjects(command.groupName, duplicate.groupName);
	XCTAssertEqualObjects(command.lPubCommandString, duplicate.lPubCommandString);
	XCTAssertEqualObjects(command.commandString, duplicate.commandString);
	XCTAssertNotIdentical(command, duplicate);
}


- (void)test_LPubRemoveGroup_CopiedObjectHasNonIdenticalValues
{
	LPubRemoveGroup *command = [[LPubRemoveGroup alloc] init];
	command.groupName = @"Some string";

	LPubRemoveGroup *duplicate = [command copyWithZone:nil];
	command.groupName = @"Another string";
	
	XCTAssertNotEqualObjects(command.groupName, duplicate.groupName);
	XCTAssertNotEqualObjects(command.lPubCommandString, duplicate.lPubCommandString);
	XCTAssertNotEqualObjects(command.commandString, duplicate.commandString);
}


- (void)test_LPubRemoveGroup_IfParametersIsLessThan3_NoLPubCommandCreated
{
	NSArray<NSString *> *parameters = @[@"REMOVE", @"GROUP"];
	
	LPubCommand *command = [LPubRemoveGroup lpubCommandInstance:parameters];
	
	XCTAssertNil(command);
}


- (void)test_LPubRemoveGroup_IfParametersIsMoreThan3_NoLPubCommandCreated
{
	NSArray<NSString *> *parameters = @[@"REMOVE", @"GROUP", @"\"name\"", @"extra"];
	
	LPubCommand *command = [LPubRemoveGroup lpubCommandInstance:parameters];
	
	XCTAssertNil(command);
}


- (void)test_LPubRemoveGroup_IfWrongParameters_NoLPubCommandCreated
{
	NSArray<NSString *> *parameters = @[@"REMOVE", @"GROUPS", @"\"name\""];
	
	LPubCommand *command = [LPubRemoveGroup lpubCommandInstance:parameters];
	
	XCTAssertNil(command);
}


- (void)test_LPubRemoveGroup_IfExpectedParameters_ProperCommandCreated
{
	NSArray<NSString *> *parameters = @[@"REMOVE", @"GROUP", @"\"name\""];
	
	LPubCommand *command = [LPubRemoveGroup lpubCommandInstance:parameters];
	
	XCTAssertIdentical(command.class, LPubRemoveGroup.class);
	XCTAssertEqualObjects(((LPubRemoveGroup *)command).groupName, @"name");
	XCTAssertEqualObjects(command.lPubCommandString, @"REMOVE GROUP \"name\"");
	XCTAssertEqualObjects(command.commandString, @"!LPUB REMOVE GROUP \"name\"");
}


- (void)test_LPubRemoveGroup_finishParsing_DoesNothing
{
	NSArray<NSString *> *parameters = @[@"REMOVE", @"GROUP", @"\"name\""];
	LPubRemoveGroup *command = (LPubRemoveGroup *)[LPubRemoveGroup lpubCommandInstance:parameters];
	LPubRemoveGroup *duplicate = [command copy];

	MockScanner *scanner = [MockScanner new];
	[command finishParsing:(id)scanner];
	
	XCTAssertEqualObjects(command.groupName, duplicate.groupName);
}


- (void)test_LPubRemoveGroup_groupName_AssignedAsCopy
{
	NSArray<NSString *> *parameters = @[@"REMOVE", @"GROUP", @"\"name\""];
	LPubRemoveGroup *command = (LPubRemoveGroup *)[LPubRemoveGroup lpubCommandInstance:parameters];

	NSMutableString *originalString = [@"group" mutableCopy];
	command.groupName = originalString;
	[originalString appendString:@"2"];
	
	XCTAssertEqualObjects(originalString, @"group2");
	XCTAssertEqualObjects(command.groupName, @"group");
}

@end

File: trunk/Packages/LDrawCore/Tests/LDrawCoreTests/LPubRemoveGroupTests.swift
//
//  LPubRemoveGroupTests.swift
//  LDrawCoreTests
//
//  How 0 !LPUB REMOVE GROUP is parsed, archived and copied.
//
//  Created by Sergey Slobodenyuk on 2023-03-09.
//

import Testing
import Foundation
import LDrawCore

@Suite("LPub REMOVE GROUP")
struct LPubRemoveGroupTests {

    private func parsed(_ parameters: [String]) -> LPubRemoveGroup? {
        perform("lpubCommandInstance:", on: LPubRemoveGroup.self, with: parameters)
    }

    @Test("The name, the LPub text and the whole line are archived, and they read back")
    func archiveRoundTrip() throws {
        let encoder = MockArchiver()
        let command = LPubRemoveGroup()
        command.groupName = "Some string"

        command.encode(with: encoder)

        #expect(encoder.data.count == 3)
        #expect(encoder.data["groupName"] as? String == "Some string")
        #expect(encoder.data["lpubCommandString"] as? String == "REMOVE GROUP \"Some string\"")
        #expect(encoder.data["commandString"] as? String == "!LPUB REMOVE GROUP \"Some string\"")

        let decoded = try #require(LPubRemoveGroup(coder: encoder))
        #expect(decoded.groupName == "Some string")
        #expect(decoded.lPubCommandString == "REMOVE GROUP \"Some string\"")
        #expect(decoded.commandString == "!LPUB REMOVE GROUP \"Some string\"")
    }

    @Test("A copy has the same name and text and is a separate object")
    func copyIsSeparate() throws {
        let command = LPubRemoveGroup()
        command.groupName = "Some string"

        let duplicate = try #require(command.copy() as? LPubRemoveGroup)
        #expect(duplicate.groupName == "Some string")
        #expect(duplicate.lPubCommandString == command.lPubCommandString)
        #expect(duplicate.commandString == command.commandString)
        #expect(duplicate !== command)

        command.groupName = "Another string"
        #expect(duplicate.groupName == "Some string")
        #expect(duplicate.lPubCommandString != command.lPubCommandString)
        #expect(duplicate.commandString != command.commandString)
    }

    @Test("Anything but REMOVE GROUP and one name is not this command", arguments: [
        ["REMOVE", "GROUP"],
        ["REMOVE", "GROUP", "\"name\"", "extra"],
        ["REMOVE", "GROUPS", "\"name\""],
    ])
    func wrongWordsGiveNothing(parameters: [String]) {
        #expect(parsed(parameters) == nil)
    }

    @Test("REMOVE GROUP and a quoted name makes the command")
    func expectedWordsMakeTheCommand() throws {
        let command = try #require(parsed(["REMOVE", "GROUP", "\"name\""]))

        #expect(ObjectIdentifier(type(of: command)) == ObjectIdentifier(LPubRemoveGroup.self))
        #expect(command.groupName == "name")
        #expect(command.lPubCommandString == "REMOVE GROUP \"name\"")
        #expect(command.commandString == "!LPUB REMOVE GROUP \"name\"")
    }

    /// The name has a space, so a plain split on spaces would lose it.
    @Test("Retyped text is read again, quoted name and all")
    func retypedTextIsReadAgain() throws {
        let command = try #require(parsed(["REMOVE", "GROUP", "\"name\""]))

        command.lPubCommandString = "REMOVE GROUP \"foo bar\""
        #expect(command.groupName == "foo bar")
        #expect(command.lPubCommandString == "REMOVE GROUP \"foo bar\"")

        command.lPubCommandString = "REMOVE GROUP \"wheels\""
        #expect(command.groupName == "wheels")
    }

    @Test("Finishing a parse leaves the name alone")
    func finishParsingChangesNothing() throws {
        let command = try #require(parsed(["REMOVE", "GROUP", "\"name\""]))

        command.finishParsing(Scanner(string: ""))

        #expect(command.groupName == "name")
    }

    @Test("The group name is copied, not shared")
    func nameIsCopied() throws {
        let command = try #require(parsed(["REMOVE", "GROUP", "\"name\""]))
        let original = NSMutableString(string: "group")

        command.groupName = original as String
        original.append("2")

        #expect(command.groupName == "group")
    }
}
